### ISSUE TITLE
Unclosed files

### ISSUE THREAD

**Issue by enfeizhan** [Theano/Theano#5261]
Got a lot of warnings saying files unclosed.

> 

Using gpu device 0: GeForce GT 650M (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 2000.)
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/cmodule.py:2182: ResourceWarning: unclosed file <_io.BufferedWriter name='/dev/null'>
  p_out = output_subprocess_Popen(cmd)
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/cmodule.py:2182: ResourceWarning: unclosed file <_io.BufferedWriter name='/dev/null'>
  p_out = output_subprocess_Popen(cmd)
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/cmodule.py:2182: ResourceWarning: unclosed file <_io.BufferedWriter name='/dev/null'>
  p_out = output_subprocess_Popen(cmd)
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/cmodule.py:2182: ResourceWarning: unclosed file <_io.BufferedWriter name='/dev/null'>
  p_out = output_subprocess_Popen(cmd)
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/cmodule.py:2182: ResourceWarning: unclosed file <_io.BufferedWriter name='/dev/null'>
  p_out = output_subprocess_Popen(cmd)
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/cmodule.py:2182: ResourceWarning: unclosed file <_io.BufferedWriter name='/dev/null'>
  p_out = output_subprocess_Popen(cmd)
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/cmodule.py:2182: ResourceWarning: unclosed file <_io.BufferedWriter name='/dev/null'>
  p_out = output_subprocess_Popen(cmd)
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/cmodule.py:2182: ResourceWarning: unclosed file <_io.BufferedWriter name='/dev/null'>
  p_out = output_subprocess_Popen(cmd)
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/cmodule.py:2182: ResourceWarning: unclosed file <_io.BufferedWriter name='/dev/null'>
  p_out = output_subprocess_Popen(cmd)
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/gof/utils.py:497: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/feizhan/miniconda/envs/py3_env/lib/python3.4/site-packages/theano/sandbox/cuda/cuda_ndarray.cuh'>
  return hash_from_code(open(file_path, 'rb').read())
>
For instance, 
```
return hash_from_code(open(file_path, 'rb').read())
```

Perhaps this better should be replaced with

```
with open(file_path, 'rb') as f:
    file_content = f.read()
return hash_from_code(file_content)
```

Printing out these warnings severely slows down the speed.

**Comment by enfeizhan** [comment]
BTW, what is the requirement for CUDA and cuDNN? Looked around but couldn't find the specs. Just install the latest ones would be OK?

**Comment by nouiz** [comment]
Theano support the latest cudnn 5.1 and cuda 8.0.

make sure to use Theano dev version. Some of those problem are fixed or
will be fixed shortly in a PR. But not all of them.

**Comment by enfeizhan** [comment]
This is supposed to be a minor issue. I can fix it if I got a chance. @nouiz

**Comment by nouiz** [comment]
Go for it! Thanks.

**Comment by enfeizhan** [comment]
hi @nouiz, I've modified the code as the example in the first post. However I have no idea of what caused issues around ``output_subprocess_Popen(cmd)`` in ``gof/cmodule.py``. 

I have created a pull request.

**Comment by lamblin** [comment]
Closed by #5307.
